### PR TITLE
Feat/unified invoices

### LINKED ISSUE #542: c-lightning should support on-chain fallback for BOLT11 invoices
Maybe via an additional arg to the invoice RPC?

### PR DESCRIPTION
**Pull request by chrisguida**
Support for tracking payments to fallback addresses (fixes #542)

Currently implemented behind the `unified_invoices` config option.

When this option is enabled, populate new invoices with a P2TR fallback address, and add the script to our `txfilter` so we can watch for payments to it, and store it in the new `invoice_fallbacks` child table. Adding fallbacks manually in the `invoice` rpc call will disable the new tracking behavior, and there is a warning to this effect.

When we detect a payment to one of our fallback addresses, check the `invoice_fallbacks` table to see if there's an invoice associated with that script. If so, mark the invoice as paid and send out any hooks / notifications for this.

TODO:

- write tests
- handle edge cases

### CODE REVIEW

**Comment by rustyrussell** [comment]
Hmm,. this is too late foir current release, but can we get that bolt11 bugfix in a  PR by itself (ideally with a Changelog-Fixed: line?)

**Comment by niftynei** [comment]
@rustyrussell done, see  #6435

File: lightningd/invoice.c [hunk 1]
```diff
@@ -1136,6 +1146,12 @@ static struct command_result *json_invoice(struct command *cmd,
 	}
 
 	if (fallbacks) {
+		if (cmd->ld->unified_invoices) {
```
Comment (rustyrussell) [inline lightningd/invoice.c:1209]: This is backwards?

File: wallet/db.c [hunk 2]
```diff
@@ -200,6 +200,12 @@ static struct migration dbmigrations[] = {
 	 "  UNIQUE (payment_hash)"
 	 ");"),
      NULL},
+    {SQL("CREATE TABLE invoice_fallbacks ("
+	 "  scriptpubkey BLOB,"
+	 "  invoice_id BIGINT REFERENCES invoices(id) ON DELETE CASCADE,"
+	 "  PRIMARY KEY (scriptpubkey)"
+	 ");"),
+     NULL},
```
Comment (rustyrussell) [inline wallet/db.c]: This *MUST* go at the end of the dbmigrations array!!

File: lightningd/chaintopology.c [hunk 3]
```diff
@@ -86,10 +88,22 @@ static void filter_block_txs(struct chain_topology *topo, struct block *b)
...
 		if (txfilter_match(topo->bitcoind->ld->owned_txfilter, tx)) {
+			// fprintf(stderr, "CAG txfilter_match found \n");
+			fprintf(stderr, "CAG txfilter_match found %s\n", fmt_bitcoin_txid(tmpctx, &txid));
+			size_t k;
 			wallet_extract_owned_outputs(topo->bitcoind->ld->wallet,
 						     tx->wtx, is_coinbase, &b->height, &owned);
 			wallet_transaction_add(topo->ld->wallet, tx->wtx,
 					       b->height, i);
+			// invoice_check_onchain_payment(tx);
+			for (k = 0; k < tx->wtx->num_outputs; k++) {
```
Comment (rustyrussell) [inline lightningd/chaintopology.c:99]: declare k inside loop, as "for (size_t k = 0;..."

File: lightningd/chaintopology.c [hunk 4]
```diff
@@ -86,10 +88,22 @@ static void filter_block_txs(struct chain_topology *topo, struct block *b)
 		owned = AMOUNT_SAT(0);
 		txid = b->txids[i];
 		if (txfilter_match(topo->bitcoind->ld->owned_txfilter, tx)) {
```
Comment (rustyrussell) [inline lightningd/chaintopology.c:90]: Maybe we should clean this up to return an array of output numbers, rather than scanning multiple times as we do...

File: lightningd/chaintopology.c [hunk 5]
```diff
@@ -86,10 +88,22 @@ static void filter_block_txs(struct chain_topology *topo, struct block *b)
...
+			fprintf(stderr, "CAG txfilter_match found %s\n", fmt_bitcoin_txid(tmpctx, &txid));
+			size_t k;
 			wallet_extract_owned_outputs(topo->bitcoind->ld->wallet,
 						     tx->wtx, is_coinbase, &b->height, &owned);
 			wallet_transaction_add(topo->ld->wallet, tx->wtx,
 					       b->height, i);
+			// invoice_check_onchain_payment(tx);
+			for (k = 0; k < tx->wtx->num_outputs; k++) {
+			const u8 *oscript = bitcoin_tx_output_get_script(tmpctx, tx, k);
+				if (txfilter_scriptpubkey_matches(topo->bitcoind->ld->owned_txfilter, oscript)) {
```
Comment (rustyrussell) [inline lightningd/chaintopology.c:101]: Don't need to check this twice...

File: lightningd/invoice.c [hunk 6]
```diff
@@ -1168,6 +1197,7 @@ static struct command_result *json_invoice(struct command *cmd,
 
 		p2tr = scriptpubkey_p2tr(fallback_scripts, &pubkey);
 		fallback_scripts[0] = p2tr;
+
```
Comment (rustyrussell) [inline lightningd/invoice.c]: remove whitespace change

File: wallet/invoices.c [hunk 7]
```diff
@@ -374,6 +375,31 @@ bool invoices_find_by_rhash(struct invoices *invoices,
 	}
 }
 
+bool invoices_find_by_fallback_script(struct invoices *invoices,
+			    u64 *inv_dbid,
```
Comment (rustyrussell) [inline wallet/invoices.c:430]: Generally prefer returned values last, so scriptPubkey before inv_dbid...

File: wallet/invoices.c [hunk 8]
```diff
@@ -374,6 +375,31 @@ bool invoices_find_by_rhash(struct invoices *invoices,
...
+{
+	struct db_stmt *stmt;
+
+	fprintf(stderr, "CAG invoices_find_by_fallback_script\n");
+
+	stmt = db_prepare_v2(invoices->wallet->db, SQL("SELECT invoice_id"
+					       "  FROM invoice_fallbacks"
+					       " WHERE scriptpubkey = ?;"));
+	db_bind_blob(stmt, scriptPubkey,
+			  tal_bytelen(scriptPubkey));
```
Comment (rustyrussell) [inline wallet/invoices.c:441]: db_bind_talarr is a convenience helper for this...

File: wallet/invoices.c [hunk 9]
```diff
@@ -375,6 +375,42 @@ bool invoices_find_by_rhash(struct invoices *invoices,
...
+	stmt = db_prepare_v2(
+	    invoices->wallet->db,
+	    SQL("INSERT INTO invoice_fallbacks"
+		"            ( invoice_id, scriptpubkey )"
+		"     VALUES ( ?, ?);"));
+
+	fprintf(stderr, "CAG invoices_find_by_fallback_script\n");
+
+	db_bind_u64(stmt, inv_dbid);
+	db_bind_blob(stmt, scriptPubkey,
```
Comment (rustyrussell) [inline wallet/invoices.c:411]: db_bind_talarr

File: wallet/invoices.c [hunk 10]
```diff
@@ -375,6 +375,42 @@ bool invoices_find_by_rhash(struct invoices *invoices,
 	}
 }
 
+bool invoices_create_fallback(struct invoices *invoices,
```
Comment (rustyrussell) [inline wallet/invoices.c:393]: void, as it cannot fail

File: lightningd/chaintopology.c [hunk 11]
```diff
@@ -92,15 +92,17 @@ static void filter_block_txs(struct chain_topology *topo, struct block *b)
 			fprintf(stderr, "CAG txfilter_match found %s\n", fmt_bitcoin_txid(tmpctx, &txid));
 			size_t k;
 			wallet_extract_owned_outputs(topo->bitcoind->ld->wallet,
-						     tx->wtx, is_coinbase, &b->height, &owned);
+										 tx->wtx, is_coinbase, &b->height, &owned);
```
Comment (rustyrussell) [inline lightningd/chaintopology.c:95]: Weird whitespace change...

File: lightningd/invoice.c [hunk 12]
```diff
@@ -470,18 +474,21 @@ invoice_check_payment(const tal_t *ctx,
 
 void invoice_try_pay(struct lightningd *ld,
 		     struct htlc_set *set,
-		     const struct invoice_details *details)
+		     const struct invoice_details *details,
+			 struct amount_msat msat)
```
Comment (rustyrussell) [inline lightningd/invoice.c]: weird indent

File: lightningd/notification.c [hunk 13]
```diff
@@ -161,12 +161,14 @@ void notify_warning(struct lightningd *ld, struct log_entry *l)
...
 						   struct amount_msat amount,
 						   struct preimage preimage,
+						   struct bitcoin_outpoint outpoint,
 						   const struct json_escape *label)
 {
 	json_object_start(stream, "invoice_payment");
 	json_add_string(stream, "msat",
 			type_to_string(tmpctx, struct amount_msat, &amount));
 	json_add_hex(stream, "preimage", &preimage, sizeof(preimage));
+	json_add_outpoint(stream, "outpoint", &outpoint);
```
Comment (rustyrussell) [inline lightningd/notification.c:171]: iff it's non-NULL...

File: lightningd/notification.c [hunk 14]
```diff
@@ -175,16 +177,17 @@ REGISTER_NOTIFICATION(invoice_payment,
...
 void notify_invoice_payment(struct lightningd *ld, struct amount_msat amount,
-			    struct preimage preimage, const struct json_escape *label)
+			    struct preimage preimage, const struct json_escape *label, struct bitcoin_outpoint *outpoint)
 {
 	void (*serialize)(struct json_stream *,
 			  struct amount_msat,
 			  struct preimage,
-			  const struct json_escape *) = invoice_payment_notification_gen.serialize;
+			  const struct json_escape *,
+			  struct bitcoin_outpoint *) = invoice_payment_notification_gen.serialize;
```
Comment (rustyrussell) [inline lightningd/notification.c:186]: parameter order here differs from actual function...

File: lightningd/chaintopology.c [hunk 15]
```diff
@@ -100,11 +100,12 @@ static void filter_block_txs(struct chain_topology *topo, struct block *b)
...
-					struct bitcoin_outpoint outpoint;
-					outpoint.txid = txid;
-					outpoint.n = k;
+					struct bitcoin_outpoint *outpoint = tal(tmpctx, struct bitcoin_outpoint);
+					outpoint->txid = txid;
+					outpoint->n = k;
 					fprintf(stderr, "CAG txfilter_scriptpubkey_matches found for script %s\n", tal_hex(tmpctx, oscript));
 					bitcoin_tx_output_get_amount_sat(tx, k, &amount);
+					// fprintf(stderr, "CAG checking outpoint: %s\n", type_to_string(tmpctx, struct bitcoin_outpoint, outpoint));
 					invoice_check_onchain_payment(topo->ld, oscript, amount, outpoint);
```
Comment (rustyrussell) [inline lightningd/chaintopology.c]: This change seems wrong...  Did you just want to hand &outpoint?

File: lightningd/invoice.c [hunk 16]
```diff
@@ -60,6 +60,12 @@ static void json_add_invoice_fields(struct json_stream *response,
 		json_add_amount_msat(response,
 				     "amount_received_msat", inv->received);
 		json_add_u64(response, "paid_at", inv->paid_timestamp);
+		if (inv->paid_outpoint) {
+			json_object_start(response, "paid_outpoint");
+			json_add_txid(response, "paid_txid", &inv->paid_outpoint->txid);
+			json_add_num(response, "paid_outnum", inv->paid_outpoint->n);
```
Comment (rustyrussell) [inline lightningd/invoice.c:66]: `paid_` in subfields is redundant

File: lightningd/invoice.c [hunk 17]
```diff
@@ -481,6 +488,7 @@ void invoice_try_pay(struct lightningd *ld,
 			 struct amount_msat msat,
 			 struct bitcoin_outpoint *outpoint)
```
Comment (rustyrussell) [inline lightningd/invoice.c:489]: const struct bitcoin_outpoint *

File: lightningd/invoice.c [hunk 18]
```diff
@@ -489,7 +497,7 @@ void invoice_try_pay(struct lightningd *ld,
 	payload->preimage = details->r;
 	payload->msat = msat;
 	payload->set = set;
-	payload->outpoint = outpoint;
+	payload->outpoint = tal_steal(payload, outpoint);
```
Comment (rustyrussell) [inline lightningd/invoice.c:500]: Ah!  If you want your own lifetimed copy, use `tal_dup_or_null(payload, struct outpoint, outpoint)`

File: lightningd/invoice.c [hunk 19]
```diff
@@ -976,15 +984,15 @@ static void listincoming_done(const char *buffer,
 }
 
 void invoice_check_onchain_payment(struct lightningd *ld,
-					  const u8 *scriptPubKey, struct amount_sat sat, struct bitcoin_outpoint outpoint)
+					  const u8 *scriptPubKey, struct amount_sat sat, struct bitcoin_outpoint *outpoint STEALS)
```
Comment (rustyrussell) [inline lightningd/invoice.c:987]: STEALS is always ugly, and in this case it's better for this to make a copy if it wants.  Also, `const struct...`

File: wallet/invoices.c [hunk 20]
```diff
@@ -96,10 +97,22 @@ static struct invoice_details *wallet_stmt2invoice_details(const tal_t *ctx,
...
 		dtl->paid_timestamp = db_col_u64(stmt, "paid_timestamp");
+		if (!db_col_is_null(stmt, "paid_txid")) {
+			paid_outpoint = tal(ctx, struct bitcoin_outpoint);
+			db_col_txid(stmt, "paid_txid",
+					&paid_outpoint->txid);
+			paid_outpoint->n
+				= db_col_int(stmt, "paid_outnum");
+			dtl->paid_outpoint = paid_outpoint;
+		} else {
+			dtl->paid_outpoint = NULL;
```
Comment (rustyrussell) [inline wallet/invoices.c:110]: db_col_ignore(stmt, "paid_outnum");

Comment (rustyrussell) [reply wallet/invoices.c:110]: You don't need to ignore paid_txid, you tested it...

File: doc/lightning-listconfigs.7.md [hunk 21]
```diff
@@ -305,6 +305,7 @@ On success, an object is returned, containing:
 - **experimental-peer-storage** (boolean, optional): `experimental-peer-storage` field from config or cmdline, or default **deprecated, removal in v24.05** *(added v23.02)*
 - **experimental-quiesce** (boolean, optional): `experimental-quiesce` field from config or cmdline, or default **deprecated, removal in v24.05** *(added v23.08)*
 - **experimental-upgrade-protocol** (boolean, optional): `experimental-upgrade-protocol` field from config or cmdline, or default **deprecated, removal in v24.05** *(added v23.08)*
+- **unified-invoices** (boolean, optional): `unified-invoices` field from config or cmdline, or default *(added v23.11)*
```
Comment (rustyrussell) [inline doc/lightning-listconfigs.7.md:308]: Should we call it `invoices-onchain-fallback`?  And I think it should merely be the default: we should be able to set it in a per-invoice basis as well.  Though that's a bit trickier; allowing `true` in fallbacks might break GRPC...

**Review by rustyrussell — COMMENTED** [review]
This looks great!  Complaints are in form, not function!

- Commits need some cleaning up
- Remove all the //

File: lightningd/chaintopology.c (hunk 5 above)
Comment (chrisguida) [reply lightningd/chaintopology.c:101]: What do you mean?

First we check to see if the tx matches. Then we have to find which outpoint inside the tx is ours (txfilter_scriptpubkey_matches is a new function that does this) so we can check to see if the amount is sufficient to satisfy the invoice (and mark which outpoint was used, for our records). Probably could be cleaned up a bit, but I'm not immediately seeing a place where we check anything twice. It's early though, so perhaps I missed something :)

File: lightningd/notification.c (hunk 14 above)
Comment (chrisguida) [reply lightningd/notification.c:186]: Whoops, thanks for catching this!

File: lightningd/chaintopology.c (hunk 15 above)
Comment (chrisguida) [reply lightningd/chaintopology.c]: IIRC the outpoint was getting dropped somewhere between here and when the notification is sent, so I needed to allocate it. Does handing `&outpoint` accomplish this as well?

File: lightningd/chaintopology.c (hunk 5 above)
Comment (chrisguida) [reply lightningd/chaintopology.c:101]: If you mean that we extract our owned outputs in `wallet_extract_owned_outputs`, yes, that is true. I needed to decide whether to alter `wallet_extract_owned_outputs` to return the owned outpoints or just scan the tx again, processing each owned outpoint as it's found, and the latter seemed cleaner. Lmk if you want it the other way.

File: wallet/invoices.c (hunk 20 above)
Comment (chrisguida) [reply wallet/invoices.c:110]: Ok, thanks!

File: doc/lightning-listconfigs.7.md (hunk 21 above)
Comment (chrisguida) [reply doc/lightning-listconfigs.7.md:308]: Sure, sounds good.

Currently if you want to avoid using this fallback-tracking functionality you can pass in a fallback manually; this will result in the original behavior (and will also generate a warning).

File: wallet/db.c (hunk 2 above)
Comment (chrisguida) [reply wallet/db.c]: Yes, this was moved to the bottom in a later commit.

**Comment by rustyrussell** [comment]
Rebased and cleaned up:
1. Renamed option.
2. Documented it in lightningd-config.md
3. Fixed serialization order
4. Documented new notification field.
5. Added Changelog lines
6. Removed many // and fprintf.

**Comment by chrisguida** [comment]
Looks great @rustyrussell, thanks again for the assist on this!

Looks like there are just a couple of things left:

1) the rest of the review comments
2) renaming the rest of the `unified_invoices` references to `invoices_onchain_fallback`
3) making the new option default to true

I'll get started on this unless I hear back!

**Comment by rustyrussell** [comment]
I think I got them all, except for minor internal code things which can be fixed later...

**Comment by rustyrussell** [comment]
YA rebase!

**Comment by rustyrussell** [comment]
We can make it per-invoice next release, if we want, since it would be perfectly backward compatible.  For now it's getting close to release...
